base/special/gamma.jl

```diff
@@ -630,12 +607,98 @@ function eta(z::Union{Float64,Complex{Float64}})
 
-"""
-    eta(x)
 
-Dirichlet eta function ``\\eta(s) = \\sum^\\infty_{n=1}(-1)^{n-1}/n^{s}``.
-"""
-eta(z::Complex) = oftype(float(z),eta(Complex128(z)))
+# Converting types that we can convert, and not ones we can not
+# If we really cared about single precision, we could make a faster
+# Float32 version by truncating the Stirling series at a smaller cutoff.
+# and if we really cared about half precision, we could make a faster
+# Float16 version, by using a precomputed table look-up.
+
+# Float16, and Float32 and their Complex equivalents can be converted to Float64
+# and results converted back. Similar for BitIntegers (eg Int32), but no converting back
+# Otherwise, we need to make things use their own `float` converting methods
+# and in those cases, we do not convert back either as we assume
+# they also implement there own versions of the functions
+
+
+ofpromotedtype(as::Tuple, c) = convert(promote_type(typeof.(as)...), c)
+ComplexOrRealUnion(TS...) = Union{(ComplexOrReal{T} for T in TS)...}
+
+const types_le_Float64 = (Float16, Float32, Float64, Base.BitInteger.types...)
+
+for T in types_le_Float64
+    @eval f64(x::Complex{$T}) = Complex128(x)
+    @eval f64(x::$T) = Float64(x)
+end
+
+
+for f in (:digamma, :trigamma, :zeta, :eta, :invdigamma)
+    @eval begin
+        $f(z::ComplexOrRealUnion(Base.BitInteger.types...)) = $f(f64(z))
+        $f(z::ComplexOrRealUnion(Float16,Float32)) = oftype(z, $f(f64(z)))
+
+        function $f(z::Number)
+            x = float(z)
+            typeof(x) === typeof(z) && throw(MethodError($f, (z,)))
+            # There is nothing to fallback to, since this didn't change the argument types
+            $f(x)
+        end
+    end
+end
+
+function polygamma(m::Integer, z::ComplexOrRealUnion(Float16,Float32))
+    oftype(z, polygamma(m, f64(z)))
+end
+
+for T1 in types_le_Float64, T2 in types_le_Float64
+    if (T1 == T2 == Float64) ||  (T1 == Int && T2 == Float64)
+        continue # Avoid redefining base definition
+        # However this skips `zeta(::Complex{Int}, ::ComplexOrReal{Float64})`
+        # so that will need to be added back after
+    end
+
+    if T1<:Integer && T2<:Integer
+        @eval function zeta(s::ComplexOrReal{$T1}, z::ComplexOrReal{$T2})
+            zeta(f64(s), f64(z)) #Do not promote down to Integers
+        end
+    else
+        @eval function zeta(s::ComplexOrReal{$T1}, z::ComplexOrReal{$T2})
+            ofpromotedtype((s, z), zeta(f64(s), f64(z)))
+        end
+    end
+end
+
+# this is the one definition that is skipped
+function zeta(s::Complex{Int}, z::ComplexOrReal{Float64})::Complex{Float64}
+    zeta(f64(s), f64(z))
+end
+
+function zeta(s::Integer, z::Number)
+    t = Int(s)  # One could worry here about converting a BigInteger into a Int32/Int64
+    x = float(z)
+    if typeof(t) === typeof(s) && typeof(x) === typeof(z)
+        # There is nothing to fallback to, since this didn't work
+        throw(MethodError(zeta,(s,z)))
+    end
```

Comment:
I think this should have a second fallback to `zeta(::Number, Number)` maybe.
Or possibly to prevent looping just try and convert `s` to a float and see how that goes.